File: org.glite.lb.server/src/seqcode.c
#ident "$Header$"
/*
Copyright (c) Members of the EGEE Collaboration. 2004-2010.
See http://www.eu-egee.org/partners for details on the copyright holders.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/


#include <errno.h>

#include "glite/lbu/trio.h"
#include "glite/lb/context-int.h"

#include "get_events.h"
#include "store.h"
#include "jobstat.h"
/*
#include "lb_authz.h"
*/
#include "db_supp.h"
#include "lb_proto.h"


int edg_wll_QuerySequenceCodeServer(edg_wll_Context ctx, edg_wlc_JobId jobid, const char *source, char **seqcode)
{
	glite_lbu_Statement	sh;
	intJobStat	   *istat = NULL;
	char		   *jobid_md5 = NULL,
				   *stmt = NULL,
				   *res = NULL,
				   *res_rest = NULL;
	int				nstates;


	edg_wll_ResetError(ctx);
	jobid_md5 = edg_wlc_JobIdGetUnique(jobid);

	trio_asprintf(&stmt,
				"select int_status from states "
				"where jobid='%|Ss' and version='%|Ss'",
				jobid_md5, INTSTAT_VERSION);

	if ( stmt == NULL ) return edg_wll_SetError(ctx, ENOMEM, NULL);

	if ( (nstates = edg_wll_ExecSQL(ctx, stmt, &sh)) < 0 ) goto cleanup;
	if ( nstates == 0 ) {
		edg_wll_SetError(ctx, ENOENT, "no state in DB");
		goto cleanup;
	}
	if ( edg_wll_FetchRow(ctx, sh, 1, NULL, &res) < 0 ) goto cleanup;

	istat = dec_intJobStat(res, &res_rest);
	if ( res_rest  && istat ) {
		*seqcode = istat->last_seqcode;
		istat->last_seqcode = NULL;
	}
	else edg_wll_SetError(ctx, EDG_WLL_ERROR_DB_CALL,
						"error decoding DB intJobStatus");

cleanup:
	free(res);
	free(jobid_md5);
	free(stmt);
	glite_lbu_FreeStmt(&sh);
	if ( istat ) {
		destroy_intJobStat(istat);
		free(istat);
	}

	return edg_wll_Error(ctx, NULL, NULL);
}

File: org.glite.lb.client-java/src/org/glite/lb/SeqCode.java
/*
Copyright (c) Members of the EGEE Collaboration. 2004-2010.
See http://www.eu-egee.org/partners for details on the copyright holders.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package org.glite.lb;

/**
 * This class represents sequence code.
 * 
 * @author Pavel Piskac (173297@mail.muni.cz)
 * @version 9. 4. 2008
 */
public class SeqCode {

    public static final int NORMAL = 1;
    public static final int DUPLICATE = 11;
    public static final int PBS = 2;
    public static final int CONDOR = 4;
    public static final int CREAM = 4;
    public static final int CREAMWMS = 5;
    
    private int[] seqCode = {0, 0, 0, 0, 0, 0, 0, 0, 0};
    private int type = 0;
    private String cream_code;
    
    /**
     * Empty constructor which creates new instance of SeqCode with all values
     * equal 0
     */
    public SeqCode() {
    }
    
    /**
     * Constructor which creates new instance of SeqCode with values set by user
     * in input attribute
     *
     * @param type type of sequence code
     * @param seqCodeString sequence code string
     */
    public SeqCode(int type,String seqCodeString) {
	getSeqCodeFromString(type,seqCodeString);
    }

    /**
     * This method increments one specific part of sequence code given by part attribute
     * 
     * @param src the source, whose part of sequence number will be increased
     */
    public void incrementSeqCode(Sources src) {
	int part = src.source;
	switch (type) {
		case NORMAL:
		case DUPLICATE:
		        if (src == null)
				throw new IllegalArgumentException("SeqCode part");
       	 		seqCode[part-1]++;
			break;
		case CREAMWMS:
			if (src == null)
                                throw new IllegalArgumentException("SeqCode part");
			seqCode[src.LRMS-1] += 1000; //XXX hardcoded to add 1000 to LRMS when WMS jobs goes through CREAM
			break;
		default: break;
       	 }
    }
    
    /**
     * Converts string representation of sequence code to format which is used
     * in this class.
     * Insert sequence codes in format:
     * UI=000000:NS=0000000000:WM=000000:BH=0000000000:JSS=000000:LM=000000:LRMS=000000:APP=000000:LBS=000000
     * @param type type of sequence code (NORMAL, DUPLICATE, CREAMWMS, CREAM)
     * @param seqCodeString sequence code string
     */
    public void getSeqCodeFromString(int type,String seqCodeString) {
	switch (type) {
		case NORMAL:
		case DUPLICATE:
		case CREAMWMS:
			if (!seqCodeString.matches("UI=\\d{1,}:NS=\\d{1,}:WM=\\d{1,}:BH=\\d{1,}:" +
			"JSS=\\d{1,}:LM=\\d{1,}:LRMS=\\d{1,}:APP=\\d{1,}:LBS=\\d{1,}")) {
			throw new IllegalArgumentException("this is not correct sequence code: " + seqCodeString);
			}
			
			int currentPosition = 0;
			int equalsPosition = 0;
			int colonPosition = 0;
			for (int i = 0; i <= 8; i++) {
				equalsPosition = seqCodeString.indexOf('=', currentPosition);
				if (i == 8) {
				colonPosition = seqCodeString.length();
				} else {
				colonPosition = seqCodeString.indexOf(':', currentPosition);
				}
				seqCode[i] = (new Integer(seqCodeString.substring(equalsPosition+1, colonPosition))).intValue();
				currentPosition = colonPosition + 1;
			}
			break;
		case CREAM: 
			cream_code = seqCodeString;
			break;
		default: throw new IllegalArgumentException("unsupported seqcode type " + type);
	}
	this.type = type;
    }
    
    public String toString() {        
	switch (type) {
		case NORMAL:
		case DUPLICATE:
		case CREAMWMS:
			String tmp = Integer.toString(seqCode[0]);    
			String output = "UI=";
			output += "000000".substring(0, 6 - tmp.length ()) + tmp;
			output += ":";
			output += "NS=";
			tmp = Integer.toString(seqCode[1]);
			output += "0000000000".substring(0, 10 - tmp.length ()) + tmp;
			output += ":";
			output += "WM=";
			tmp = Integer.toString(seqCode[2]);
			output += "0000000000".substring(0, 6 - tmp.length ()) + tmp;
			output += ":";
			output += "BH=";
			tmp = Integer.toString(seqCode[3]);
			output += "0000000000".substring(0, 10 - tmp.length ()) + tmp;
			output += ":";
			output += "JSS=";
			tmp = Integer.toString(seqCode[4]);
			output += "0000000000".substring(0, 6 - tmp.length ()) + tmp;
			output += ":";
			output += "LM=";
			tmp = Integer.toString(seqCode[5]);
			output += "0000000000".substring(0, 6 - tmp.length ()) + tmp;
			output += ":";
			output += "LRMS=";
			tmp = Integer.toString(seqCode[6]);
			output += "0000000000".substring(0, 6 - tmp.length ()) + tmp;
			output += ":";
			output += "APP=";
			tmp = Integer.toString(seqCode[7]);
			output += "0000000000".substring(0, 6 - tmp.length ()) + tmp;
			output += ":";
			output += "LBS=";
			tmp = Integer.toString(seqCode[8]);
			output += "0000000000".substring(0, 6 - tmp.length ()) + tmp;
			return output;
		case CREAM: 
			return cream_code;
		default:
			throw new IllegalArgumentException("unitialized seqcode");
	}
    }
}
